### DISCUSSION TITLE
How can I declare additionalProperties: {} on a component?

### DISCUSSION THREAD

**Discussion by Ryan-DL** [juhaku/utoipa#247] [Q&A]
I need to declare a free form object similar to this on a type on a schema. (https://swagger.io/docs/specification/data-models/data-types/ example stolen from the docs)

```
type: object
additionalProperties: {}
```

How can I declare this in a struct? I was thinking something like this.

```
#[derive(Debug, Serialize, Deserialize, Component)]
pub struct SomeComponent {
    #[component(value_type = Object)]
    pub values: Value,
}
```

Thanks in advance 👍

**Comment by juhaku** [comment]
Hi,

Currently in latest release 1.1.0 there is basically no support for this. But in master branch there is support for this which can be then achieved as follows:
```rust
#[derive(Debug, Serialize, Deserialize, ToSchema)]
pub struct SomeComponent {
    #[schema(value_type = Any)]
    pub values: Value,
}
```
This will be available in new 2.0.0 version once released. The syntax `#[schema(value_type = Any)]` makes the value behave like an arbitrary object rendering it as `type: object` within the the OpenAPI spec.

In current release 1.1.0 the `value_type` does only support primitive type conversions.

**Reply by juhaku** [reply]
The `additionalProperties: {}` will only be used when the type rendered is a map to the item type of the map entry and cannot be defined currently manually.

**Reply by willbuckner** [reply]
Hmm, how do we specifically require that it's an object though? `Any` implied any type, for example, an `int` would be valid. One idea might be to at least allow manually specifying the OpenAPI spec type string. As it is, our docs are incorrect without being able to require an object specifically.

**Reply by juhaku** [reply]
Well the idea behind the `Any` is exactly that. Any is anything thus not any specifically declared type. 

> One idea might be to at least allow manually specifying the OpenAPI spec type string

I'm not a fan for this. Stringly typed APIs are bad and exactly the reason why we use Rust so we can leverage the undisputed type system it offers. If we would offer users to write the type in string form it would allow users to specify pretty much anything even incorrect value. Unless the content of the string was validated but how it would then be different from the type format of the possible value.

However I changed the `Any` to an `Object` which would follow the OpenAPI spec more closely. https://github.com/juhaku/utoipa/pull/256
